File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char

data QuestionType = Question
                  | Yell
                  | Empty
                  | Default

responseFor :: String -> String
responseFor xs = case inputType xs of
                      Question -> "Sure."
                      Yell -> "Whoa, chill out!"
                      Empty -> "Fine. Be that way!"
                      Default -> "Whatever."


inputType :: String -> QuestionType
inputType string | all isSpace sanitisedString = Empty
                 | not (any isLower sanitisedString) && any isAlpha sanitisedString = Yell
                 | (last sanitisedString) == '?' = Question
                 | otherwise = Default
  where sanitisedString = filter (/= ' ') string

File: lua/bob/bob.lua
local m = {}
local tests = {
  function(str) -- Check if string is empty
    return (#str == 0 or nil) and "Fine, be that way."
  end,
  function(str) -- Check if is a question (ends with '?')
    return string.match(str, "%?$") and "Sure"
  end,
  function(str) -- Check if yelling (upper == original)
    return (string.upper(str) == str or nil) and "Whoa, chill out!"
  end
}
local dont_understand = "Whatever"

function m.hey(bob_str)
  for i,test in ipairs(tests) do
    local test_result = test(bob_str)
    if test_result ~= nil then
      return test_result
    end
  end
  return dont_understand
end

return m
